### DISCUSSION TITLE
Using RgbColor::Compare to detect end point for color blending fade.

### DISCUSSION THREAD

**Discussion by thatguy-jaysenodell** [Makuna/NeoPixelBus#814] [Q&A] [answered]
I'm building a small, multi-strip controller based on the atmega328p (arduino uno if it was still an uno). The NeoPixelBus library has made things simple but I like to do most things "the hard way" because I'm stupid. So instead of using animations, I wanted to make my own "fade between color" routine. I'm fighting with RgbColor::Compare to determine when my updates have reached the end state. 

quick notes:
- I have several "chambers" with strips in them, each with multiple default colors. _currRgbColor is the current active "full strip" color. _alert is a bool stating that this is exception state 
- This is just dirty code for now to play with the ide. that stupid print was trying to understand what Compare was returning
- Feel free to point out some obvious thing I'm missing in this. 
- I get stuck looping with that println ALWAYS reporting 230 irrespective of what values I supply for my right and left

Here is my code:
```
void Chamber::setLevel(RgbColor oLvl){
	RgbColor color = RgbColor::LinearBlend(_currRgbColor, oLvl, 0.1f);

	while (RgbColor::Compare( color, oLvl, 200) != 0) {
		Serial.println(RgbColor::Compare( color, oLvl, 200));
		RgbColor color = RgbColor::LinearBlend(color, oLvl, 0.1f);
		for (int i=0; i< _strip->Pixels(); i++){
			_strip->SetPixelColor(i, color);
		}
		_strip->Show();
		delay(5);
	}
	if( !_alert ) { _currRgbColor = color; }
}
```

My assumption is that I'm using the RgbColor::Compare incorrectly (or missing somthing obvious). I could not see any use of RgbColor::Compare in the examples. 

Thank you in advance for any hints.

**Comment by Makuna** [comment, marked as answer]
https://github.com/Makuna/NeoPixelBus/wiki/RgbColor-object-API#static-int16_t-compareconst-rgbcolor-left-const-rgbcolor-right-uint8_t-epsilon--1
Epsilon is how sloppy you want the compare to be.  You provided 200!  Usually, 1, the default is good.  Sometimes depending on how you are blending, you might go as high 6.  But anything over 25 you may need to rethink your approach.

I think your iterative approach isn't a good one.  Think time, how long do you want your blend action to take, from one color to another color.  The LinearBlend can take either 0-255 or 0.0 -1.0 for progress.  Iterate on this progress value from the original color to the new color, and not rely on comparing to end the loop.  Then change the time based on the steps you take for progress and the delay you use in the loop.  This is the technique used in the animation examples, look at the animation callback.  All the animation system really does is manage the time and this progress value as a float (0.0-1.0) for multiple "animations" so reduce code that you would need to implement.

**Reply by thatguy-jaysenodell** [reply]
Just to clarify your suggested approach, if I'm considering a 100 loop fade then the better approach for the looping sections would be 
```
for (int i = 0; i < 100; i++;){
  RgbColor newColor = RgbColor::LinearBlend(_currRgbColor, oLvl, i);
  for (int q=0; q < _strip->Pixels(); q++){
    _strip->SetPixelColor(q, newColor);
  }
  _strip.Show();
  delay(5);
}

```
Attempting to confirm my comprehending of the exampels as well as your reply.

**Reply by thatguy-jaysenodell** [reply]
edit because testing shows you have to go to 255 to get all the way to the end and shortening the delay vs using the map function may be cheaper

**Reply by Makuna** [reply]
You are calling the wrong method to get the count of pixels.  
[Pixels()](https://github.com/Makuna/NeoPixelBus/wiki/NeoPixelBus-object-API#size_t-pixelssize) returns a pointer to the bytes of raw data.
[PixelCount()](https://github.com/Makuna/NeoPixelBus/wiki/NeoPixelBus-object-API#uint16_t-pixelcount) returns the number of pixels.

You can also just map the 0-100 to 0-255, by using `i * 255 / 99`

**Reply by Makuna** [reply]
also, you can set all the pixels to a color by calling [ClearTo()](https://github.com/Makuna/NeoPixelBus/wiki/NeoPixelBus-object-API#void-cleartocolorobject-color)

**Reply by thatguy-jaysenodell** [reply]
Ha! I guess that's a testament to your code that it's working very well with my stupid in it. Even dumber is that I have it stored in a static variable in my Chamber object. But...

I was starting with the count method as I'm planning to have a 6x24 matix that I will need to light in zones. in playing with topologies I foind it difficult to define a square from 2,4 to 4,6 and then turn it white. my plan was to eventually turn my method into a "do it the hard way" generic "change color of random pix" by feeding it maps from config. I'm sure you have a way to do this, I just haven't looked hard enough at it as I've been fighting with simple things and kicked the can down the road for now. That said I see that ClearTo() has a start and stop so if I can get the points in each matrix segment then this still works for my generic method.

Thank you for your help with this. Let me make a couple changes and get back to you. As of now, my original quesion has been answered with your original pointer to use the measured progress method.

**Reply by thatguy-jaysenodell** [reply]
To complete the answer with corrected code, here are the resulting functions in my class that accomplish stated goals using @Makuna suggestions. everythign works and the results in smaller than my original code. 

**notes**:
- _fadeSpeed is set in class to allow adusting the rate of fade (set to 254 when no fade, set to 1 when slow fade). 
- the initial setLevel(int oLvl) allows for reading from analog read to set ambient levels (used with door switches and ntp server to adjust levels)
- _setLevel(RgbColor) is called by a few other class routines (for now) and still needs extended to manage my "bright box". but this is much simpler for now.

```
void Chamber::setLevel(int oLvl){
        // map the analog signal to the limits we set for non-alert brightnesss for this chamber
        // send that as a white level to setLevel for standard
	setLevel( RgbColor( map(oLvl, 0, 1024, _minNormalBright, _maxNormalBright) ) );
}

void Chamber::setLevel(RgbColor oLvl){
        // run through 255 possible itterations of color blend
        // use _fadeSpeed to jump itterations
	for ( uint8_t q = 0; q <= 254; q += _fadeSpeed ){
		_setLevel(RgbColor::LinearBlend( _currRgbColor, oLvl, q));
	}
       // when going to black fast we tend to not get to the last step to oLvl once to force it
	_setLevel(oLvl);

        // store oLvl as current string color
	if( !_alert ) { _currRgbColor = oLvl; }
}
void Chamber::_setLevel(RgbColor color){
	// we don't need a for loop to hit each pix.. who knew?
        // that would be folks to read ALL the API docs. 
        _strip->ClearTo(color);
	_strip->Show();
}
```
